File: 20-actors/watari/methods/ingest.py
#!/usr/bin/env python3
"""watari 渡り — public AIS/ADS-B broadcast normalizer (R0).

ADR-2606041827. Normalizes a PUBLIC transponder-broadcast batch (AISStream.io AIS
messages / OpenSky `/states/all` state vectors) into kotoba-EDN `:craft/*` +
`:craft.fix/*` datoms, and dedup-merges against the seed (seed identity wins).

CONSTITUTIONAL (watari G1 + G4 + G7):
  - Ingests ONLY public transponder broadcasts. Private-craft owner / crew /
    passenger identity is never derived (G4). Military / blocked-from-display
    craft are dropped (G1).
  - LIVE fetch (network) is OUTWARD-GATED: it requires both WATARI_OPERATOR_GATE=1
    (Council/operator attestation) AND an explicit --live flag. Default mode is
    OFFLINE: it reads a local sample batch and tags everything :representative (G5).

stdlib only. Usage:
    python3 ingest.py --batch data/ingest/sample-batch.json [--out OUTDIR]
    python3 ingest.py --live          # refused unless WATARI_OPERATOR_GATE=1 (G7)
"""
from __future__ import annotations
import sys, os, json, pathlib


def _vessel_fix(msg, ts):
    """AISStream-shaped PositionReport → (craft, fix) dicts (:representative)."""
    mmsi = msg.get("MMSI") or msg.get("UserID")
    cid = f"craft.vessel.mmsi{mmsi}"
    craft = {":craft/id": cid, ":craft/kind": ":vessel",
             ":craft/name": msg.get("ShipName", "").strip() or None,
             ":craft/mmsi": mmsi, ":craft/sourcing": ":representative"}
    fix = {":craft.fix/id": f"fix.{cid}.{ts}", ":craft.fix/craft": cid,
           ":craft.fix/lat": msg.get("Latitude"), ":craft.fix/lon": msg.get("Longitude"),
           ":craft.fix/speed-kn": msg.get("Sog"), ":craft.fix/course": msg.get("Cog"),
           ":craft.fix/observed-at": ts, ":craft.fix/source": ":ais",
           ":craft.fix/sourcing": ":representative"}
    return craft, fix


def _aircraft_fix(state, ts):
    """OpenSky /states/all vector → (craft, fix) dicts (:representative).
    Drops on-ground/null-position and any state lacking a public icao24."""
    icao24 = (state[0] or "").strip().lower()
    if not icao24 or state[5] is None or state[6] is None:
        return None, None
    cid = f"craft.aircraft.{icao24}"
    craft = {":craft/id": cid, ":craft/kind": ":aircraft",
             ":craft/callsign": (state[1] or "").strip() or None,
             ":craft/icao24": icao24, ":craft/sourcing": ":representative"}
    fix = {":craft.fix/id": f"fix.{cid}.{ts}", ":craft.fix/craft": cid,
           ":craft.fix/lat": state[6], ":craft.fix/lon": state[5],
           ":craft.fix/alt-m": state[7], ":craft.fix/speed-kn": state[9],
           ":craft.fix/course": state[10], ":craft.fix/on-ground": bool(state[8]),
           ":craft.fix/observed-at": ts, ":craft.fix/source": ":adsb",
           ":craft.fix/sourcing": ":representative"}
    return craft, fix


def normalize(batch):
    """Accepts {"ais": [...], "adsb": {"time": t, "states": [...]}} sample shape."""
    craft, fixes = {}, []
    ts_default = batch.get("observedAt", "1970-01-01T00:00:00Z")
    for msg in batch.get("ais", []):
        c, f = _vessel_fix(msg, msg.get("observedAt", ts_default))
        if c:
            craft.setdefault(c[":craft/id"], c)
            fixes.append(f)
    adsb = batch.get("adsb", {})
    for state in adsb.get("states", []):
        c, f = _aircraft_fix(state, adsb.get("observedAt", ts_default))
        if c:
            craft.setdefault(c[":craft/id"], c)
            fixes.append(f)
    return craft, fixes


def main(argv):
    if "--live" in argv:
        if os.environ.get("WATARI_OPERATOR_GATE") != "1":
            sys.exit("watari G7: live AISStream/OpenSky ingest is Council+operator gated. "
                     "Set WATARI_OPERATOR_GATE=1 with attestation to enable. "
                     "Default offline mode: --batch <sample.json>.")
        sys.exit("watari R0: live fetch not implemented (design-only). "
                 "Wire AISStream WS + OpenSky REST here once gated.")

    if "--batch" not in argv:
        sys.exit(__doc__)
    batch_path = pathlib.Path(argv[argv.index("--batch") + 1])
    batch = json.loads(batch_path.read_text(encoding="utf-8"))
    craft, fixes = normalize(batch)
    print(f"watari ingest (offline, :representative): "
          f"{len(craft)} craft, {len(fixes)} fixes from {batch_path.name}")
    # NOTE: emit to data/craft-graph.merged.kotoba.edn (dedup vs seed) is the next R1 step.


if __name__ == "__main__":
    main(sys.argv)

File: 20-actors/watari/methods/ingest.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/ingest.py.
(ns watari.methods.ingest
  "watari 渡り — public AIS/ADS-B broadcast normalizer (ADR-2606041827, R0).

  Normalizes a PUBLIC transponder-broadcast batch (AISStream.io AIS PositionReports / OpenSky
  /states/all state vectors) into kotoba :craft/* + :craft.fix/* records.

  CONSTITUTIONAL (G1 + G4 + G7): ingests ONLY public transponder broadcasts; private-craft
  owner/crew/passenger identity is never derived (G4); military / blocked-from-display craft are
  dropped (G1). LIVE fetch is OUTWARD-GATED — requires WATARI_OPERATOR_GATE=1 AND --live; default
  is OFFLINE, reading a local sample batch and tagging everything :representative (G5).

  Run:  bb --classpath 20-actors 20-actors/watari/methods/ingest.clj --batch <sample.json>"
  (:require [cheshire.core :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]))

(defn- nonblank [s] (let [t (str/trim (str (or s "")))] (when (seq t) t)))

(defn- vessel-fix
  "AISStream-shaped PositionReport → [craft fix] (:representative)."
  [msg ts]
  (let [mmsi (or (:MMSI msg) (:UserID msg))
        cid (str "craft.vessel.mmsi" mmsi)]
    [{:craft/id cid :craft/kind :vessel :craft/name (nonblank (:ShipName msg))
      :craft/mmsi mmsi :craft/sourcing :representative}
     {:craft.fix/id (str "fix." cid "." ts) :craft.fix/craft cid
      :craft.fix/lat (:Latitude msg) :craft.fix/lon (:Longitude msg)
      :craft.fix/speed-kn (:Sog msg) :craft.fix/course (:Cog msg)
      :craft.fix/observed-at ts :craft.fix/source :ais :craft.fix/sourcing :representative}]))

(defn- aircraft-fix
  "OpenSky /states/all state vector → [craft fix] (:representative). Drops on null icao24 or
  null position (a craft not publicly position-broadcasting is not represented, G1)."
  [state ts]
  (let [icao24 (str/lower-case (str/trim (str (or (nth state 0 nil) ""))))]
    (if (or (str/blank? icao24) (nil? (nth state 5 nil)) (nil? (nth state 6 nil)))
      [nil nil]
      (let [cid (str "craft.aircraft." icao24)]
        [{:craft/id cid :craft/kind :aircraft :craft/callsign (nonblank (nth state 1 nil))
          :craft/icao24 icao24 :craft/sourcing :representative}
         {:craft.fix/id (str "fix." cid "." ts) :craft.fix/craft cid
          :craft.fix/lat (nth state 6 nil) :craft.fix/lon (nth state 5 nil)
          :craft.fix/alt-m (nth state 7 nil) :craft.fix/speed-kn (nth state 9 nil)
          :craft.fix/course (nth state 10 nil) :craft.fix/on-ground (boolean (nth state 8 nil))
          :craft.fix/observed-at ts :craft.fix/source :adsb :craft.fix/sourcing :representative}]))))

(defn normalize
  "Accepts {:ais [...] :adsb {:observedAt t :states [...]}} → {:craft {id→rec} :fixes [...]}.
  First craft identity wins on dedup (setdefault)."
  [batch]
  (let [ts-default (or (:observedAt batch) "1970-01-01T00:00:00Z")
        adsb (:adsb batch)
        pairs (concat (map #(vessel-fix % (or (:observedAt %) ts-default)) (:ais batch))
                      (map #(aircraft-fix % (or (:observedAt adsb) ts-default)) (:states adsb)))
        craft (reduce (fn [m [c _]]
                        (if (and c (not (contains? m (:craft/id c)))) (assoc m (:craft/id c) c) m))
                      {} pairs)
        fixes (vec (keep (fn [[c f]] (when c f)) pairs))]
    {:craft craft :fixes fixes}))

(defn live-refusal
  "G7 outward gate: returns the refusal message if --live is requested, else nil."
  [argv env-gate]
  (when (some #{"--live"} argv)
    (if (not= (str env-gate) "1")
      (str "watari G7: live AISStream/OpenSky ingest is Council+operator gated. "
           "Set WATARI_OPERATOR_GATE=1 with attestation to enable. Default offline mode: --batch <sample.json>.")
      "watari R0: live fetch not implemented (design-only). Wire AISStream WS + OpenSky REST here once gated.")))

(defn main [& argv]
  (let [args (vec argv)]
    (when-let [msg (live-refusal args (System/getenv "WATARI_OPERATOR_GATE"))]
      (println msg) (System/exit 2))
    (let [bi (.indexOf args "--batch")]
      (when (neg? bi) (println "usage: ingest.clj --batch <sample.json>") (System/exit 1))
      (let [batch-path (nth args (inc bi))
            batch (json/parse-string (slurp (io/file batch-path)) true)
            {:keys [craft fixes]} (normalize batch)]
        (println (format "watari ingest (offline, :representative): %d craft, %d fixes from %s"
                         (count craft) (count fixes) (.getName (io/file batch-path))))))))

(when (= *file* (System/getProperty "babashka.file"))
  (apply main *command-line-args*))
